File: clipy_macos/src/gui/design_tokens.rs
use gpui::{Pixels, Size, px, size};

/// 与 Swift `DesignTokens.swift` / `AppWindowSize` 对齐
pub mod spacing {
    use gpui::Pixels;
    use gpui::px;

    pub fn xs() -> Pixels {
        px(8.0)
    }
    pub fn sm() -> Pixels {
        px(12.0)
    }
    pub fn md() -> Pixels {
        px(16.0)
    }
    pub fn lg() -> Pixels {
        px(20.0)
    }
    pub fn xl() -> Pixels {
        px(28.0)
    }
}

pub mod font {
    pub const CAPTION: f32 = 11.0;
    pub const BODY: f32 = 13.0;
    pub const SECONDARY: f32 = 12.0;
    pub const TITLE: f32 = 14.0;
    pub const EMPTY_STATE: f32 = 16.0;
}

pub mod row {
    use gpui::Pixels;
    use gpui::px;

    pub fn compact() -> Pixels {
        px(28.0)
    }
    pub fn standard() -> Pixels {
        px(36.0)
    }
    pub fn group() -> Pixels {
        px(40.0)
    }
}

pub mod radius {
    use gpui::Pixels;
    use gpui::px;

    pub fn sm() -> Pixels {
        px(4.0)
    }
    pub fn md() -> Pixels {
        px(8.0)
    }
    pub fn badge() -> Pixels {
        px(10.0)
    }
}

pub mod chrome {
    use gpui::Pixels;
    use gpui::px;

    pub fn toolbar() -> Pixels {
        px(40.0)
    }
    pub fn tab_bar() -> Pixels {
        px(36.0)
    }
    pub fn status() -> Pixels {
        px(28.0)
    }
    pub fn sidebar() -> Pixels {
        px(240.0)
    }
}

pub mod window_size {
    use super::*;

    pub fn settings() -> Size<Pixels> {
        size(px(420.0), px(640.0))
    }
    pub fn list() -> Size<Pixels> {
        size(px(720.0), px(500.0))
    }
    pub fn search() -> Size<Pixels> {
        size(px(1200.0), px(800.0))
    }
    pub fn editor() -> Size<Pixels> {
        size(px(800.0), px(600.0))
    }
    pub fn log() -> Size<Pixels> {
        size(px(800.0), px(500.0))
    }
    pub fn search_min() -> Size<Pixels> {
        size(px(800.0), px(680.0))
    }
    pub fn editor_min() -> Size<Pixels> {
        size(px(640.0), px(480.0))
    }
    pub fn list_min() -> Size<Pixels> {
        size(px(480.0), px(320.0))
    }
    pub fn notification_min() -> Size<Pixels> {
        size(px(560.0), px(360.0))
    }
    pub fn preview_min_width() -> Pixels {
        px(280.0)
    }
    pub fn table_min_width() -> Pixels {
        px(420.0)
    }
}

File: clipy_macos/Sources/UI/DesignTokens.swift
import AppKit
import SwiftUI

enum AppSpacing {
    static let xxs: CGFloat = 4
    static let xs: CGFloat = 8
    static let sm: CGFloat = 12
    static let md: CGFloat = 16
    static let lg: CGFloat = 20
    static let xl: CGFloat = 24
    static let section: CGFloat = 28
}

enum AppFont {
    static let captionSize: CGFloat = 11
    static let bodySize: CGFloat = 13
    static let secondarySize: CGFloat = 12
    static let emptyStateSize: CGFloat = 16

    static var caption: Font { .system(size: captionSize) }
    static var body: Font { .system(size: bodySize) }
    static var secondary: Font { .system(size: secondarySize) }
    static var emptyState: Font { .system(size: emptyStateSize) }
    static var title: Font { .system(size: 20, weight: .semibold) }
    static var section: Font { .system(size: 13, weight: .semibold) }
}

// MARK: - Crash-safe AppKit text attributes

extension AppFont {
    /// AppKit's font factories are imported as non-optional but can still hand back nil —
    /// `monospacedSystemFont` resolves the hidden `.AppleSystemUIFontMonospaced` asset and
    /// fails on some machines. Swift lets that nil into an attribute dictionary and CoreText
    /// aborts the process while laying out the string, so route every dictionary through here.
    static func textAttributes(
        size: CGFloat,
        weight: NSFont.Weight = .regular,
        monospaced: Bool = false,
        color: NSColor?
    ) -> [NSAttributedString.Key: Any] {
        var attributes: [NSAttributedString.Key: Any] = [:]
        if let font = resolveFont(size: size, weight: weight, monospaced: monospaced) {
            attributes[.font] = font
        }
        if let color {
            attributes[.foregroundColor] = color
        }
        return attributes
    }

    /// Adds `.font`/`.foregroundColor` only when they actually resolved.
    static func attributes(font: NSFont?, color: NSColor?) -> [NSAttributedString.Key: Any] {
        var attributes: [NSAttributedString.Key: Any] = [:]
        if let font { attributes[.font] = font }
        if let color { attributes[.foregroundColor] = color }
        return attributes
    }

    static func resolveFont(size: CGFloat, weight: NSFont.Weight = .regular, monospaced: Bool = false) -> NSFont? {
        if monospaced {
            return firstResolved(
                // Proportional metrics with tabular digits — same alignment benefit for
                // numeric readouts without depending on the monospaced font asset.
                NSFont.monospacedDigitSystemFont(ofSize: size, weight: weight),
                NSFont.monospacedSystemFont(ofSize: size, weight: weight),
                NSFont(name: "Menlo", size: size),
                NSFont.systemFont(ofSize: size)
            )
        }
        return firstResolved(
            NSFont.systemFont(ofSize: size, weight: weight),
            NSFont.systemFont(ofSize: size),
            NSFont(name: "Helvetica", size: size)
        )
    }

    private static func firstResolved(_ candidates: NSFont?...) -> NSFont? {
        candidates.first { $0 != nil } ?? nil
    }
}

enum AppRowHeight {
    static let compact: CGFloat = 30
    static let standard: CGFloat = 40
    static let group: CGFloat = 44
}

enum AppCornerRadius {
    static let small: CGFloat = 6
    static let medium: CGFloat = 8
    static let large: CGFloat = 12
    static let badge: CGFloat = 10
}

enum ScreenshotChrome {
    static let toolbarHeight: CGFloat = 44
    static let barHeight: CGFloat = 40
    static let secondaryBarHeight: CGFloat = 32
    static let floatingRadius: CGFloat = 12
    static let magnifierSize: CGFloat = 120
    static let snapThreshold: CGFloat = 8
    static let presetColors: [NSColor] = [.systemRed, .systemOrange, .systemYellow, .systemGreen, .systemBlue]
}

enum AppWindowSize {
    static let settings = CGSize(width: 760, height: 620)
    static let settingsMin = CGSize(width: 660, height: 480)
    static let screenshotSettings = CGSize(width: 800, height: 680)
    static let list = CGSize(width: 720, height: 500)
    static let search = CGSize(width: 1080, height: 720)
    static let editor = CGSize(width: 1120, height: 720)
    static let log = CGSize(width: 800, height: 500)
    static let passwordGenerator = CGSize(width: 460, height: 520)
    static let passwordGeneratorMin = CGSize(width: 420, height: 470)
    static let listMin = CGSize(width: 480, height: 320)
    static let searchMin = CGSize(width: 800, height: 520)
    static let notificationMin = CGSize(width: 560, height: 360)
    static let editorMin = CGSize(width: 920, height: 540)
}
